crates/red_knot_python_semantic/src/types/protocol_class.rs

```diff
@@ -0,0 +1,230 @@
+        self.is_protocol(db).then_some(ProtocolClassLiteral(self))
+    }
+}
+
+/// Representation of a single `Protocol` class definition.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(super) struct ProtocolClassLiteral<'db>(ClassLiteral<'db>);
+
+impl<'db> ProtocolClassLiteral<'db> {
+    /// Returns the protocol members of this class.
+    ///
+    /// A protocol's members define the interface declared by the protocol.
+    /// They therefore determine how the protocol should behave with regards to
+    /// assignability and subtyping.
+    ///
+    /// The list of members consists of all bindings and declarations that take place
+    /// in the protocol's class body, except for a list of excluded attributes which should
+    /// not be taken into account. (This list includes `__init__` and `__new__`, which can
+    /// legally be defined on protocol classes but do not constitute protocol members.)
+    ///
+    /// It is illegal for a protocol class to have any instance attributes that are not declared
+    /// in the protocol's class body. If any are assigned to, they are not taken into account in
+    /// the protocol's list of members.
+    pub(super) fn interface(self, db: &'db dyn Db) -> &'db ProtocolInterface<'db> {
+        let _span = tracing::trace_span!("protocol_members", "class='{}'", self.name(db)).entered();
+        cached_protocol_interface(db, *self)
+    }
+
+    pub(super) fn is_runtime_checkable(self, db: &'db dyn Db) -> bool {
+        self.known_function_decorators(db)
+            .contains(&KnownFunction::RuntimeCheckable)
+    }
+}
+
+impl<'db> Deref for ProtocolClassLiteral<'db> {
+    type Target = ClassLiteral<'db>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// The interface of a protocol: the members of that protocol, and the types of those members.
+#[derive(Debug, PartialEq, Eq, salsa::Update, Default, Clone, Hash)]
+pub(super) struct ProtocolInterface<'db>(Box<[ProtocolMember<'db>]>);
+
+impl<'db> ProtocolInterface<'db> {
+    /// Iterate over the members of this protocol.
+    pub(super) fn members(&self) -> impl ExactSizeIterator<Item = &ProtocolMember<'db>> {
+        self.0.iter()
+    }
+
+    /// Return `true` if all members of this protocol are fully static.
+    pub(super) fn is_fully_static(&self, db: &'db dyn Db) -> bool {
+        self.members().all(|member| member.ty.is_fully_static(db))
+    }
+
+    /// Return `true` if if all members on `self` are also members of `other`.
+    ///
+    /// TODO: this method should consider the types of the members as well as their names.
+    /// TODO: using a map as the underlying data structure would be more efficient here than a boxed slice.
+    pub(super) fn is_sub_interface_of(&self, other: &Self) -> bool {
+        self.members().all(|member| {
+            other
+                .members()
+                .any(|other_member| member.name() == other_member.name())
+        })
+    }
```

Comment:
I'd like for `ProtocolInterface` to wrap an `FxOrderMap<Name, ProtocolMemberData<'db>>` rather than a `Box<[ProtocolMember<'db>]>` -- then this wouldn't need to be `O(n^2)` in complexity. But `FxOrderMap` needs to have `salsa::Update` derived for it, which can only be done from the Salsa repo due to the Orphan rule for traits (thanks @MichaReiser for helping me debug that earlier!)

Reply:
Could we use a `BTreeMap` isntead of an `FxOrderedMap`. I think salsa implements `Update` for those and `BTreeMap` has nice `difference` methods which I think `OrderedMap` doesn't have

Reply:
Oh, and we want it to be sorted anyway, so a `BTreeMap` makes a lot of sense. Of course, thanks!

Reply:
> Unless we have operations where we need to get an element by name.

we do, and we'll have more in the future